File: q51-75/q56.py
# 2020.08.24
# Problem Statememt:
# https://leetcode.com/problems/merge-intervals/

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        # check corner cases
        if len(intervals) <= 1: return intervals
        
        # build a dictionary to finally sort them by decending interval length
        dictionary = {}
        for i in range(0, len(intervals)):
            dictionary[tuple(intervals[i])] = intervals[i][1] - intervals[i][0]
        
        intervals = []
        for w in sorted(dictionary, key=dictionary.get, reverse=True):
            intervals.append(w)

        # initialize answer to return, first possible element would be the longest interval.
        answer = []
        answer.append([intervals[0][0], intervals[0][1]])

        # check from the second until the end
        for i in range(1, len(intervals)):
            # at most duplicate with 2 intervals,
            # initialize index with duplication as -1 and -1.
            combine_1, combine_2 = -1, -1
            
            # if the left of current <= right of another,
            # and the right of current >= left of another,
            # then the current should be merged with "another"
            
            # find the first
            for j in range(0, len(answer)):                
                if intervals[i][0] <= answer[j][1] and intervals[i][1] >= answer[j][0]:
                    combine_1 = j
                    break
            
            # find the second
            for j in range(0, len(answer)):                
                if intervals[i][0] <= answer[j][1] and intervals[i][1] >= answer[j][0]:
                    combine_2 = j
            
            # if only need to merge with one,
            # ie: first == second, merge with either:
            if combine_1 != -1 and combine_2 != -1 and combine_1 == combine_2: 
                answer[combine_1] = [min(answer[combine_1][0], intervals[i][0]), \
                                     max(answer[combine_1][1], intervals[i][1])]
            
            # if need to merge with both:
            if combine_1 != -1 and combine_2 != -1 and combine_1 != combine_2:
                answer[combine_1] = [min(answer[combine_1][0], answer[combine_2][0]), \
                                     max(answer[combine_1][1], answer[combine_2][1])]
                answer.remove(answer[combine_2])
            
            # if no need to merge at all, count as a new interval:
            if combine_1 == -1 and combine_2 == -1:
                answer.append([intervals[i][0], intervals[i][1]])
            
        return answer

File: Amazon/q56.java
// 2022.05.19 midnight
// Problem Statement:
// https://leetcode.com/problems/merge-intervals/

// idea: sort by first index, then merge 2 intervals at a time
class Solution {
    public int[] mergeTwo(int[] interval1, int[] interval2) {
        if (interval1[1]>=interval2[0]) {
            return new int [] {interval1[0], Math.max(interval1[1], interval2[1])};
        } else { // cannot merge to 1
            return new int [] {-1, -1};
        }
    }
    
    public int[][] merge(int[][] intervals) {
        if (intervals.length==1) return intervals;
        Arrays.sort(intervals, (a, b) -> Double.compare(a[0], b[0]));
        List<int[]> answer_list = new ArrayList<> ();
        int[] interval1=null;
        int[] interval2=null;
        for (int i=0; i<intervals.length; i++) {
            if (i==0) {
                interval1 = intervals[0];
            } else {
                interval2 = intervals[i];
                int[] temp = mergeTwo(interval1, interval2);
                if (temp[0]==-1 && temp[1]==-1) { // cannot merge, add interval1 into result
                    answer_list.add(interval1);
                    interval1 = interval2;
                } else {
                    interval1 = temp;
                }
            }
        }
        answer_list.add(interval1);
        int [][] answer = new int [answer_list.size()][2];
        for (int i=0; i<answer_list.size(); i++) {
            answer[i][0] = answer_list.get(i)[0];
            answer[i][1] = answer_list.get(i)[1];
        }
        return answer;
    }
}